spinalcordtoolbox/registration/algorithms.py

```diff
@@ -329,6 +337,220 @@ def register_step_label(src, dest, step, verbose=1):
+    i_max = len(list(pt_second_model.named_parameters()))
+    torchparam = pt_second_model.state_dict()
+    for k, v in torchparam.items():
+        if i < i_max:
+            torchparam[k] = weights_second_model[i]
+            i += 1
+    pt_second_model.load_state_dict(torchparam)
+    pt_second_model.to(device)
+    pt_second_model.eval()
+
+    # ---- Registration ---- #
+    logger.info("\n Predicting using first VxmDense model...")
+    moved, warp_tensor_first = pt_first_model(input_moving, input_fixed, registration=True)
+    warp_data_first = warp_tensor_first[0].permute(1, 2, 3, 0).detach().numpy()
+
+    logger.info("\n Predicting using second VxmDense model...")
+    moved_final, warp_tensor_second = pt_second_model(moved, input_fixed, registration=True)
+    warp_data_second = warp_tensor_second[0].permute(1, 2, 3, 0).detach().numpy()
+
+    # ---- Warping fields ---- #
+    # Modify the warp data so it can be used with sct_apply_transfo()
+    # (add a time dimension, change the sign of some axes and set the intent code to vector)
+    orientation_conv = "LPS"
+    fx_im_orientation = list(aff2axcodes(fx_preproc_nii.affine))
+    opposite_character = {'L': 'R', 'R': 'L',
+                          'A': 'P', 'P': 'A',
+                          'I': 'S', 'S': 'I'}
+    perm = [0, 1, 2]
+    inversion = [1, 1, 1]
+    for i, character in enumerate(orientation_conv):
+        try:
+            perm[i] = fx_im_orientation.index(character)
+        except ValueError:
+            perm[i] = fx_im_orientation.index(opposite_character[character])
+            inversion[i] = -1
+
+    # First Warping Field
+    # Add the time dimension
+    warp_data_exp_first = np.expand_dims(warp_data_first, axis=3)
+    warp_data_exp_first_copy = np.copy(warp_data_exp_first)
+    # Permute the axes if needed
+    warp_data_exp_first[..., 0] = inversion[0] * warp_data_exp_first_copy[..., perm[0]]
+    warp_data_exp_first[..., 1] = inversion[1] * warp_data_exp_first_copy[..., perm[1]]
+    warp_data_exp_first[..., 2] = inversion[2] * warp_data_exp_first_copy[..., perm[2]]
+
+    # Second Warping Field
+    # Add the time dimension
+    warp_data_exp_second = np.expand_dims(warp_data_second, axis=3)
+    warp_data_exp_second_copy = np.copy(warp_data_exp_second)
+    # Permute the axes if needed
+    warp_data_exp_second[..., 0] = inversion[0] * warp_data_exp_second_copy[..., perm[0]]
+    warp_data_exp_second[..., 1] = inversion[1] * warp_data_exp_second_copy[..., perm[1]]
+    warp_data_exp_second[..., 2] = inversion[2] * warp_data_exp_second_copy[..., perm[2]]
+
+    # Compose the two warping fields together (from the first and second registration)
+
+    # Forward registration
+    warp_data = warp_data_exp_first + warp_data_exp_second
+    warp = Nifti1Image(warp_data, fx_preproc_nii.affine)
+    # Set the intent code to vector
+    warp.header['intent_code'] = 1007
+    # Save the composed warping field [forward]
+    save(warp, fname_warp_forward)
```

Comment:
Which resources did you use to write this section? (e.g. did you use any NIfTI documentation for figuring out that `1007` is the intent code for vector arrays?)

I ask because I just want to make sure we're not missing anything for creating these NIfTI files... want to make sure we're setting everything correctly. So, it would be nice to have a link to refer to that explains what we're doing here.

Reply:
Thanks for pointing this out ! You're right it appears in the code as a black box but to decide to set the `intent_code` to `1007`, I have used the information found on [this](https://brainder.org/2012/09/23/the-nifti-file-format/#:~:text=As%20above%2C%20vector%20per%20voxel%2C%20stored%20in%20the%205th%20dimension.) website and confirmed it with the documentation found [here](http://users.bmap.ucla.edu/~mchamber/npl/nifti_8h_source.html#l00470:~:text=470%C2%A0%23define%20NIFTI_INTENT_VECTOR%20%20%20%201007%20%20%20/*%20for%20any%20other%20type%20of%20vector%20*/).

I have added these two links in the code with this [commit](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3807/commits/92a1a9084fe5a50b3b031888ca4b6ee25e85333e).

Reply:
I checked the links that you provided, and noticed that the spec says:

```
#define NIFTI_INTENT_DISPVECT  1006   /* specifically for displacements */
#define NIFTI_INTENT_VECTOR    1007   /* for any other type of vector */
```

My guess would have been that a warping field represents displacement vectors? So, I would have guessed that the intent code would be 1006. :thinking: 

Just to check, I've run `sct_register_to_template` using an ANTs algorithm (slicereg) then ran `sct_image -header` to check the metadata:

```console
joshua@tadpole:/tmp/pytest-of-joshua/pytest-1/test_sct_register_multimodal_mcurrent$ sct_image -i warp_mt02mt1.nii.gz -header

--
Spinal Cord Toolbox (git-jn/3760-add-voxelmorph-registration-1dfe5dc80beef565e77d8ea6495d23cf73e8c4b2*)

sct_image -i warp_mt02mt1.nii.gz -header
--

sizeof_hdr	348
data_type	FLOAT32
dim		[5, 40, 40, 5, 1, 3, 1, 1]
vox_units	mm
time_units	Unknown
datatype	16
nbyper		4
bitpix		32
pixdim		[-1.0, 0.84375, 0.84375, 17.0, 0.0, 0.0, 0.0, 0.0]
vox_offset	352
cal_max		0.000000
cal_min		0.000000
scl_slope	1.000000
scl_inter	0.000000
phase_dim	0
freq_dim	0
slice_dim	0
slice_name	Unknown
slice_code	0
slice_start	0
slice_end	0
slice_duration	0.000000
toffset		0.000000
intent		Vector
intent_code	1007
intent_name	
intent_p1	0.000000
intent_p2	0.000000
intent_p3	0.000000
qform_name	Scanner Anat
qform_code	1
qto_xyz:1	-0.843379 -0.000000 -0.503756 10.726137 
qto_xyz:2	-0.001874 0.841376 1.273740 -17.241285 
qto_xyz:3	-0.024932 -0.063247 16.944728 -25.065205 
qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
qform_xorient	Right-to-Left
qform_yorient	Posterior-to-Anterior
qform_zorient	Inferior-to-Superior
sform_name	Unknown
sform_code	0
sto_xyz:1	0.000000 0.000000 0.000000 0.000000 
sto_xyz:2	0.000000 0.000000 0.000000 0.000000 
sto_xyz:3	0.000000 0.000000 0.000000 0.000000 
sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
sform_xorient	Unknown
sform_yorient	Unknown
sform_zorient	Unknown
file_type	NIFTI-1+
file_code	1
descrip		
aux_file
```

Comparing to files generated by `algo=dl`:

```console
joshua@tadpole:/tmp/pytest-of-joshua/pytest-0/test_sct_register_multimodal_mcurrent$ sct_image -i warp_mt02mt1.nii.gz -header

--
Spinal Cord Toolbox (git-jn/3760-add-voxelmorph-registration-1dfe5dc80beef565e77d8ea6495d23cf73e8c4b2*)

sct_image -i warp_mt02mt1.nii.gz -header
--

sizeof_hdr	348
data_type	FLOAT32
dim		[5, 40, 40, 5, 1, 3, 1, 1]
vox_units	mm
time_units	Unknown
datatype	16
nbyper		4
bitpix		32
pixdim		[-1.0, 0.84375, 0.84375, 17.0, 0.0, 0.0, 0.0, 0.0]
vox_offset	352
cal_max		0.000000
cal_min		0.000000
scl_slope	1.000000
scl_inter	0.000000
phase_dim	0
freq_dim	0
slice_dim	0
slice_name	Unknown
slice_code	0
slice_start	0
slice_end	0
slice_duration	0.000000
toffset		0.000000
intent		Vector
intent_code	1007
intent_name	
intent_p1	0.000000
intent_p2	0.000000
intent_p3	0.000000
qform_name	Scanner Anat
qform_code	1
qto_xyz:1	-0.843379 -0.000000 -0.503756 10.726137 
qto_xyz:2	-0.001874 0.841376 1.273740 -17.241285 
qto_xyz:3	-0.024932 -0.063247 16.944728 -25.065205 
qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
qform_xorient	Right-to-Left
qform_yorient	Posterior-to-Anterior
qform_zorient	Inferior-to-Superior
sform_name	Unknown
sform_code	0
sto_xyz:1	0.000000 0.000000 0.000000 0.000000 
sto_xyz:2	0.000000 0.000000 0.000000 0.000000 
sto_xyz:3	0.000000 0.000000 0.000000 0.000000 
sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
sform_xorient	Unknown
sform_yorient	Unknown
sform_zorient	Unknown
file_type	NIFTI-1+
file_code	1
descrip		
aux_file
```

They are identical. `1007` seems to be the correct code, and there is no metadata missing! :)